### ISSUE TITLE
Upgrade to FastSurfer VINN and hires surface processing on 3T and 1.5T

### ISSUE THREAD

**Issue by m-reuter** [Deep-MI/FastSurfer#146]
## Feature Description

* Upgrade FastSurferCNN to our new voxel size independent FastSurferVINN 
* Upgrade recon-surf for hires processing

## Motivation

Currently we are still conforming everything to 1mm iso. FastSurferVINN https://www.sciencedirect.com/science/article/pii/S1053811922000623 
can take a range of also sub-mm resolutions and outperforms FastSurferCNN. In order to use the resulting high-res segmentations for surface reconstruction, also recon-surf script needs to be modified for hires processing. Potentially also the dockerfiles need to be adopted if more dependencies are needed (e.g. from FreeSurfer).

**Comment by jsteiglechner** [comment]
Do you have a date in mind when this feature should be introduced?
We have developed a model for high resolutions (0.6 mm) and high field strengths (7T and 9.4T). A performance comparison with FastSurferVINN for different resolutions (0.6mm-1.2mm) and field strengths (1.5T-9.4T) would be interesting. It would be great if we could include this analysis in our upcoming paper.
Many thanks in advance.

**Comment by m-reuter** [comment]
We have been making progress. For this feature to be added, we first needed to upgrade FreeSurfer to 7.2. An issue in FreeSurfer (global spherical rotation alignment) kept us back and we now implemented a workaround in #164  . This allows us to move forward with our minor release V1.1. Based on that we can upgrade both the segmentation (VINN) and the recon-surf pipeline #134 soon(ish). Before we release that (v2.0), we need to do some thorough testing both for 1mm and sub-mm images on the full pipeline. We will be working hard to get a release done before mid July, but if that fails, we are looking at end of August or September. 

Note, that the high-res stream will only be approved for 3T field strength. It would need to be adopted to 7T and beyond via re-training, augmentation etc. High field strength images look quite different from those at 3T, causing segmentation to fail unless the network has seen enough of those images in training.

**Comment by rcruces** [comment]
Hi, thanks for the great pipeline, I ran fastSurfer on 3T data and I'm very happy with the results, so we wanted to move forward and apply it on our new 7T acquisitions.
is this issue solved in the latest merge? can I already try it? we are very interested to implement it on our 7T pipeline. If it is solve how do I use it?

**Comment by m-reuter** [comment]
Hi rcruces, I have to disappoint you. This part is the first step towards high-res (sub-mm from around .7mm  upward) at 3T and 1.5T! It only adds support to the surface pipeline (recon-surf) and we are currently working on a similar upgrade to our network (FastSurferVINN see #168). Even once all that is completed and tested, it will only work for 3T (and 1.5T) images. 7T images are very different and the network will not work on them as it has not seen any 7T data in training.

So 7T will not be supported until we re-train with 7T ground truth. Sorry.

**Comment by rcruces** [comment]
Thanks for your answer, I'm still interested to test it with high-res data from 3T. We have 0.7 and 0.8mm isometric images. If you have any examples of how to use it that would be very useful. otherwise I'll wait until the README and everything is updated.

**Comment by rcruces** [comment]
Is the VINN ready to use? can I star some testing in our data? resolution of 0.5x0.5x0.5, 0.7x0.7x0.7 and 0.8x0.8x0.8

**Comment by m-reuter** [comment]
Hi @rcruces , 
We are still actively working on that branch. We will probably start testing next week. So, of course you can test the branch, but it is not supported and can change any time. Also be aware: 
- it is for 3T data, no 7T
- we have never used .5 data, so that could fail
- we have seen issues with low quality inputs T1 at 1mm from one site in abide with a strong bias field and lots of noise in the center of the T1.

**Comment by rcruces** [comment]
thanks for the reply, actually I would be very interested in testing it. Do you have new specifications to run it or a container?
also are the flags still the same or did something changed for the processing?

thanks

**Comment by m-reuter** [comment]
Take a look at the feature/vinn branch and its readme and --help texts. Some command line flags have changed (e.g. --device cpu instead of --no_cuda etc).  You can also build a container from it as described in the docker directory. It currently works together with FreeSurfer 7.3.2. which is what we will use in our tests. For  FS7.2 you need to switch off the version check via a flag (in case you use a native install, for the docker FS732 will be downloaded and used inside the container).

**Comment by rcruces** [comment]
ok perfect, will start some testing next week!

**Comment by rcruces** [comment]
Hello,

I built the container from `feature/vinn` and tried to run it but I got a lot of errors. Apparently it has something to do with a certificate, when it was trying to download or access a training file hosted at [b2share.fz-juelich.de](url).


This is the full error
```

/fastsurfer/FastSurferCNN /fastsurfer
python3.8 run_prediction.py --t1 /data/sub-PNC002_ses-01_space-nativepro_t1w.nii.gz --seg /output/sub-PNC002_ses-01/mri/aparc.DKTatlas+aseg.deep.mgz --conformed_name /output/sub-PNC002_ses-01/mri/orig.mgz --seg_log /output/sub-PNC002_ses-01/scripts/deep-seg.log --ckpt_sag /fastsurfer/checkpoints/FastSurferVINN_training_state_sagittal.pkl --ckpt_ax /fastsurfer/checkpoints/FastSurferVINN_training_state_axial.pkl --ckpt_cor /fastsurfer/checkpoints/FastSurferVINN_training_state_coronal.pkl --batch_size 8 --cfg_cor /fastsurfer/FastSurferCNN/config/FastSurferVINN_coronal.yaml --cfg_ax /fastsurfer/FastSurferCNN/config/FastSurferVINN_axial.yaml --cfg_sag /fastsurfer/FastSurferCNN/config/FastSurferVINN_sagittal.yaml --run_viewagg_on check --device cpu
[INFO: run_prediction.py:  363]: Checking or downloading default checkpoints ...
[INFO: checkpoint.py:  146]: Response code: 404
/venv/lib/python3.8/site-packages/urllib3/connectionpool.py:1045: InsecureRequestWarning: Unverified HTTPS request is being made to host 'b2share.fz-juelich.de'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
Traceback (most recent call last):
  File "/fastsurfer/FastSurferCNN/utils/checkpoint.py", line 144, in download_checkpoint
    response.raise_for_status()
  File "/venv/lib/python3.8/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: NOT FOUND for url: https://b2share.fz-juelich.de/api/files/a423a576-220d-47b0-9e0c-b5b32d45fc59/FastSurferVINN_training_state_axial.pkl

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run_prediction.py", line 364, in <module>
    get_checkpoints(args.ckpt_ax, args.ckpt_cor, args.ckpt_sag)
  File "/fastsurfer/FastSurferCNN/utils/checkpoint.py", line 175, in get_checkpoints
    check_and_download_ckpts(axi, url)
  File "/fastsurfer/FastSurferCNN/utils/checkpoint.py", line 166, in check_and_download_ckpts
    download_checkpoint(url, ckptname, checkpoint_path)     
  File "/fastsurfer/FastSurferCNN/utils/checkpoint.py", line 148, in download_checkpoint
    response.raise_for_status()
  File "/venv/lib/python3.8/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: NOT FOUND for url: https://b2share.fz-juelich.de/api/files/a423a576-220d-47b0-9e0c-b5b32d45fc59/FastSurferVINN_training_state_axial.pkl
 
```

**Comment by af-a** [comment]
Hello @rcruces.

The certificate-related issue should be resolved now. Could you try again?

**Comment by rcruces** [comment]
Yes it worked indeed. Thanks 
But now I had a different error.

My command:
```bash
docker run -v ${data_dir}:/data \
           -v "${SUBJECTS_DIR}":/output \
           -v "${fs_licence}":/fs_license \
           --rm fastsurfer:cpu \
           --t1 /data/sub-00_t1w.nii.gz \
           --sid sub-00 --sd /output \
           --fs_license /fs_license/license.txt \
           --parallel --threads 20 --device cpu \
           --surfreg
```
           
Error log:
```bash
Checking total volume ...
Voxel size in mm3: 1.0
Total segmentation volume in liter: 1.27
/fastsurfer
/fastsurfer/recon_surf /fastsurfer

sid sub-00
T1 /output/sub-00/mri/orig.mgz
seg /output/sub-00/mri/aparc.DKTatlas+aseg.deep.mgz

Checking Input Segmentation Quality ...
python3.8 /fastsurfer/recon_surf//quick_qc.py --seg /output/sub-00/mri/aparc.DKTatlas+aseg.deep.mgz
python3.8: can't open file '/fastsurfer/recon_surf//quick_qc.py': [Errno 2] No such file or directory
```

Also could you please change the permissions of the outputs (e.g. `umask 001` or `777`), we work in a cluster and we don't have root permissions, so is very difficult to modify this files.
`drwxr-xr-x 4 root    root  42 Nov 21 15:01 sub-00`


Additionally i noticed that the `orig.mgz` file was conformed to 1x1x1, so I guess that the VINN feature is not working?
```bash
orig.mgz

************************************************
Image name:          "orig.mgz"
************************************************
  Dimensions:        256 x 256 x 256
  Voxel size:        1 x 1 x 1
  Data strides:      [ -1 3 -2 ]
```

I build the docker container from the latest version of `feature/vinn`, when I print the help I get this at the end:
```bash
Henschel L, Conjeti S, Estrada S, Diers K, Fischl B, Reuter M, FastSurfer - A
 fast and accurate deep learning based neuroimaging pipeline, NeuroImage 219
 (2020), 117012. https://doi.org/10.1016/j.neuroimage.2020.117012
```
If this the branch is not ready to run some test I can wait a bit. Or can prove some feedback while I test ti.

best

Thank you

**Comment by m-reuter** [comment]
@rcruces  thanks for your feedback. Note, that the vinn branch is under active development still and can be broken at any time. For example it looks like something went wrong with the quck_qc which was merged yesterday. We will check that. 

Currently the highres processing needs to be switched on with the --hires flag. This behaviour is likely to change in the future (e.g. we may decide to make it the default). Also many command line flags will probably change soon.

**Comment by m-reuter** [comment]
@rcruces what command did you use to print the help? In run_fastsurfer the help should also print the other paper.

**Comment by af-a** [comment]
@rcruces 

There was indeed an error due to quick_qc.py; I have fixed that now.

If you do not want the input image conformed to 1x1x1, you need to pass in the `--hires` flag at the moment.

You should see the following at the end of the help description, which refers to the original FastSurfer and the FastSurferVINN publications:
```
REFERENCES:

If you use this for research publications, please cite:

Henschel L, Conjeti S, Estrada S, Diers K, Fischl B, Reuter M, FastSurfer - A
 fast and accurate deep learning based neuroimaging pipeline, NeuroImage 219
 (2020), 117012. https://doi.org/10.1016/j.neuroimage.2020.117012

Henschel L*, Kuegler D*, Reuter M. (*co-first). FastSurferVINN: Building
 Resolution-Independence into Deep Learning Segmentation Methods - A Solution
 for HighRes Brain MRI. NeuroImage 251 (2022), 118933. 
 http://dx.doi.org/10.1016/j.neuroimage.2022.118933
```

The VINN feature branch is a work in progress; we are currently still adding features and testing, but your feedback is welcome and appreciated.

**Comment by LeHenschel** [comment]
> Also could you please change the permissions of the outputs (e.g. umask 001 or 777), we work in a cluster and we don't have root permissions, so is very difficult to modify this files.
drwxr-xr-x 4 root root 42 Nov 21 15:01 sub-00

@rcruces, you can simply pass the desired user and group with the docker call. Just add `--user $(id -u):$(id -g)` or manually add user and group (`--user userID:groupID`). So the call would look like this:

```
docker run -v ${data_dir}:/data \
           -v "${SUBJECTS_DIR}":/output \
           -v "${fs_licence}":/fs_license \
           --user $(id -u):$(id -g) \
           --rm fastsurfer:cpu \
           --t1 /data/sub-00_t1w.nii.gz \
           --sid sub-00 --sd /output \
           --fs_license /fs_license/license.txt \
           --parallel --threads 20 --device cpu \
           --surfreg
```

**Comment by rcruces** [comment]
Hello, yes the flag `--help` printed both references. I rebuilt the container, and restar the test. The only little detail is that `--user` is not defined on `run_fastsurfer.sh` at least on the `feature/vinn`. This is the log:  
`ERROR: Flag --user unrecognized.`

**Comment by LeHenschel** [comment]
The --user flag is not a fastsurfer, but a **docker flag.** So it has do be defined before you call the container (**before fastsurfer:cpu**; see the example command above).

**Comment by melanieganz** [comment]
Dear @m-reuter, is there an update regarding the release of the upgrade to FastSurfer VINN and hires surface processing on 3T and 1.5T? We'd love to try it out.

**Comment by m-reuter** [comment]
Hi Melanie !  We are currently running the last tests, there is no plan to include any further functionality or changes (only if any bugs come up during the last test). An initial set of quick tests already passed, so we are confident that it will be released soon (meaning in 1-2 weeks) but only if nothing else pops up. Happy to hear that you want to try it out!

**Comment by m-reuter** [comment]
we released 2.0
